File: Agent/llm_client.py
#!/usr/bin/env python3
"""
Module for interfacing with the local LLM.
"""
from __future__ import annotations

import logging
import re
import sys
import requests
from typing import Any, Dict, List

from . import config

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s llm_client :: %(message)s")
log = logging.getLogger("llm_client")

def get_llm_completion(
    system_prompt: str,
    user_prompt: str,
) -> str:
    """
    Gets a completion from the OpenAI-compatible local LLM.
    """
    url = f"{config.LLM_API_BASE}/chat/completions"
    payload = {
        "model": config.LLM_MODEL_NAME,
        "temperature": config.LLM_TEMPERATURE,
        "messages": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt},
        ],
    }
    
    log.info(f"Sending completion request to {url} for model {config.LLM_MODEL_NAME}")
    try:
        headers = {
            "Authorization": f"Bearer {config.OPENROUTER_API_KEY}",
            "Content-Type": "application/json",
        }
        resp = requests.post(url, headers=headers, json=payload, timeout=config.LLM_TIMEOUT)

        resp.raise_for_status()
        data = resp.json()
        content = data["choices"][0]["message"]["content"]
        log.info("Successfully received LLM completion.")
        return content
    except requests.exceptions.RequestException as e:
        log.error(f"API call failed: {e}")
        # This is a critical failure for the planner
        return f"Error: API call failed. {e}"
    except (KeyError, IndexError) as e:
        log.error(f"Invalid response structure from LLM API: {e}")
        return f"Error: Invalid LLM API response. {e}"

def extract_yaml_block(text: str) -> str:
    """
    Pulls the *last* fenced YAML block from model output.
    This is more robust, as LLMs sometimes output text or
    a "thinking" block before the final code.
    """
    matches = re.findall(r"```(yaml|yml)?\s*(.+?)\s*```", text, flags=re.S | re.I)
    
    if not matches:
        log.warning("No YAML block found in LLM output. Returning raw text.")
        return text.strip()
    
    # Return the content of the *last* match
    log.info("Extracted last YAML block from LLM output.")
    return matches[-1][1].strip()

File: src/agent/llm_client.ts
import type { LlmCompletionOptions, LlmMessage } from "./types";

export class LlmClient {
	constructor(
		private baseUrl: string,
		private apiKey: string,
	) {}

	async complete(
		messages: LlmMessage[],
		options: LlmCompletionOptions,
	): Promise<string> {
		const payload = {
			model: options.model,
			temperature: options.temperature ?? 0.2,
			max_tokens: options.maxTokens ?? 2048,
			stream: false,
			messages,
		};

		const headers: Record<string, string> = {
			"Content-Type": "application/json",
		};

		if (this.apiKey) {
			headers.Authorization = `Bearer ${this.apiKey}`;
		}

		const response = await fetch(`${this.baseUrl}/chat/completions`, {
			method: "POST",
			headers,
			body: JSON.stringify(payload),
			signal: options.timeoutMs
				? AbortSignal.timeout(options.timeoutMs)
				: undefined,
		});

		if (!response.ok) {
			const body = await response.text();
			throw new Error(`LLM request failed: ${response.status} ${body}`);
		}

		const data = (await response.json()) as {
			choices?: Array<{ message?: { content?: string } }>;
			error?: { message?: string };
		};

		if (data.error?.message) {
			throw new Error(`LLM error: ${data.error.message}`);
		}

		const content = data.choices?.[0]?.message?.content?.trim();
		if (!content) {
			throw new Error("LLM response missing content");
		}

		return content;
	}
}
